File: leetcode/c++/Sort_Colors.cpp
class Solution {
public:
    void sortColors(int A[], int n) {
        if(n <= 1) return;
        
        int i = 0;
        int j = 0;
        int k = n-1;
        
        while(j <= k){
            if(A[j] == 0){
                swap(A[i], A[j]);
                i++;
                j++;
            }else if(A[j] == 1){
                j++;
            }else{
                swap(A[j], A[k]);
                k--;
            }
        }
    }
};

File: leetcode/java/SortColors.java
class Solution {
    public void sortColors(int[] nums) {
        if(nums == null || nums.length <= 1) return;
        int i = 0;
        int j = 0;
        int k = nums.length-1;
        while(j <= k) {
            if(nums[j] == 1) j++;
            else if(nums[j] < 1) swap(nums, j++, i++);
            else swap(nums, j, k--);
        }
    }
    
    private void swap(int[] A, int i, int j) {
        int x = A[i];
        A[i] = A[j];
        A[j] = x;
    }
}
